src/Crypto/Spake2/Group.hs

```diff
@@ -62,10 +51,86 @@ class Group group where
+  --
+  -- Note [Byte encoding in Group]
+  decodeElement :: ByteArray bytes => group -> bytes -> CryptoFailable (Element group)
+
+  -- | Size of elements, in bits
+  elementSizeBits :: group -> Int
+
+  -- | Deterministically create an arbitrary element from a seed bytestring.
+  --
+  -- __XXX__: jml would much rather this take a scalar, an element, or even an integer, rather than bytes
+  -- because bytes mean that the group instances have to know about hash algorithms and HKDF.
+  -- If the IntegerGroup class in SPAKE2 also oversized its input,
+  -- then it and the ed25519 implementation would have identical decoding.
+  arbitraryElement :: ByteArrayAccess bytes => group -> bytes -> Element group
+
+
+-- | A group where 'elementAdd' is commutative.
+--
+-- That is, where
+--
+-- prop> \x y -> elementAdd group x y == elementAdd group y x
+--
+-- This property leads to a natural \(\mathbb{Z}\)-module,
+-- where scalar multiplication is defined as repeatedly calling `elementAdd`.
+--
+-- === Definitions
+--
+-- Warning: this gets algebraic.
+--
+-- A /module/ is a ring \(R\) together with an abelian group \((G, +)\),
+-- and a new operator \(\cdot\) (i.e. scalar multiplication)
+-- such that:
+--
+-- 1. \(r \cdot (x + y) = r \cdot x + r \cdot y\)
+-- 2. \((r + s) \cdot x = r \cdot x + s \cdot x\)
+-- 3. \((rs) \cdot x = r \cdot (s \cdot x)\)
+-- 4. \(1_R \cdot x = x\)
+--
+-- for all \(x, y\) in \(G\), and \(r, s\) in \(R\),
+-- where \(1_R\) is the identity of the ring.
+--
+-- A /ring/ \(R, +, \cdot\) a set \(R\) with two operators such that:
+--
+-- 1. \(R\) is an abelian group under \(+\)
+-- 2. \(R\) is a monoid under \(\cdot\)
+-- 3. \(cdot\) is _distributive_ with respect to \(+\). That is,
+--    1. \(a \cdot (b + c) = (a \cdot b) + (a \cdot c) (left distributivity)
+--    2. \((b + c) \cdot a) = (b \cdot a) + (c \cdot a) (right distributivity)
+--
+-- Note we have to define left & right distributivity,
+-- because \(\cdot\) might not be commutative.
+--
+-- A /monoid/ is a group without the notion of inverse. See Haskell's 'Monoid' typeclass.
+--
+-- A \(\mathbb{Z}\)-module is a module where the ring \(R\)
+-- is the integers with normal addition and multiplication.
+class Group group => AbelianGroup group where
+  -- | A scalar for this group.
+  -- Mathematically equivalent to an integer,
+  -- but possibly stored differently for computational reasons.
+  type Scalar group :: *
+
   -- | Multiply an element of the group with respect to a scalar.
   --
   -- This is equivalent to adding the element to itself N times, where N is a scalar.
+  -- The default implementation does exactly that.
   scalarMultiply :: group -> Scalar group -> Element group -> Element group
+  scalarMultiply group scalar element =
+    scalarMultiply' (scalarToInteger group scalar) element
+    where
+      scalarMultiply' 0 _ = groupIdentity group
+      scalarMultiply' n x = elementAdd group x (scalarMultiply' (n - 1) x)
```

Comment:
@exarkun this might not be familiar to you.

This is a default implementation of the `scalarMultiply` method. Supplying it here means that instances (implementations) of this typeclass (interface), don't need to define their own `scalarMultiply` to fully implement the interface. If they don't provide one, they'll get this implementation.

We don't use it in our code, but I figure it's useful for illustrative purposes.

For another example, the [Haskell Report 2010](https://www.haskell.org/onlinereport/haskell2010/haskellch9.html#x16-1710009) defines `Eq` as:

```haskell
class  Eq a  where  
    (==), (/=) :: a -> a -> Bool  
 
        -- Minimal complete definition:  
        --      (==) or (/=)  
    x /= y     =  not (x == y)  
    x == y     =  not (x /= y)
```

Which is why when we define [equality for `ExtendedPoint`](https://github.com/jml/haskell-spake2/blob/master/src/Crypto/Spake2/Groups/Ed25519.hs#L189) we don't have to define `/=` (not equals—yeah, I wish it was `!=` too).